File: log-server/config_test.go
package main

import (
	"os"
	"testing"
	"time"
)

func TestDefaultConfig(t *testing.T) {
	cfg := DefaultConfig()

	if cfg.Port != 8080 {
		t.Errorf("default Port = %d; want 8080", cfg.Port)
	}
	if cfg.DataDir == "" {
		t.Error("default DataDir should not be empty")
	}
	if cfg.MaxUploadMB != 10 {
		t.Errorf("default MaxUploadMB = %d; want 10", cfg.MaxUploadMB)
	}
	if cfg.RetentionDays != 30 {
		t.Errorf("default RetentionDays = %d; want 30", cfg.RetentionDays)
	}
}

func TestConfigFromEnv(t *testing.T) {
	// Set env vars
	os.Setenv("WUWAID_PORT", "9090")
	os.Setenv("WUWAID_DATA_DIR", "/tmp/wuwaid-test-data")
	os.Setenv("WUWAID_MAX_UPLOAD_MB", "20")
	os.Setenv("WUWAID_RETENTION_DAYS", "14")
	defer func() {
		os.Unsetenv("WUWAID_PORT")
		os.Unsetenv("WUWAID_DATA_DIR")
		os.Unsetenv("WUWAID_MAX_UPLOAD_MB")
		os.Unsetenv("WUWAID_RETENTION_DAYS")
	}()

	cfg := ConfigFromEnv()

	if cfg.Port != 9090 {
		t.Errorf("env Port = %d; want 9090", cfg.Port)
	}
	if cfg.DataDir != "/tmp/wuwaid-test-data" {
		t.Errorf("env DataDir = %s; want /tmp/wuwaid-test-data", cfg.DataDir)
	}
	if cfg.MaxUploadMB != 20 {
		t.Errorf("env MaxUploadMB = %d; want 20", cfg.MaxUploadMB)
	}
	if cfg.RetentionDays != 14 {
		t.Errorf("env RetentionDays = %d; want 14", cfg.RetentionDays)
	}
}

func TestConfigFromEnvPartial(t *testing.T) {
	// Only set one var, rest should be defaults
	os.Setenv("WUWAID_PORT", "3000")
	defer os.Unsetenv("WUWAID_PORT")

	cfg := ConfigFromEnv()

	if cfg.Port != 3000 {
		t.Errorf("env Port = %d; want 3000", cfg.Port)
	}
	if cfg.MaxUploadMB != 10 {
		t.Errorf("default MaxUploadMB = %d; want 10 (not overridden)", cfg.MaxUploadMB)
	}
	if cfg.RetentionDays != 30 {
		t.Errorf("default RetentionDays = %d; want 30 (not overridden)", cfg.RetentionDays)
	}
}

func TestConfigMaxUploadBytes(t *testing.T) {
	cfg := Config{MaxUploadMB: 5}
	want := int64(5 * 1024 * 1024)
	if got := cfg.MaxUploadBytes(); got != want {
		t.Errorf("MaxUploadBytes() = %d; want %d", got, want)
	}
}

func TestConfigRetentionDuration(t *testing.T) {
	cfg := Config{RetentionDays: 7}
	want := 7 * 24 * time.Hour
	if got := cfg.RetentionDuration(); got != want {
		t.Errorf("RetentionDuration() = %v; want %v", got, want)
	}
}

func TestConfigLogDir(t *testing.T) {
	cfg := Config{DataDir: "/data"}
	want := "/data/logs"
	if got := cfg.LogDir(); got != want {
		t.Errorf("LogDir() = %s; want %s", got, want)
	}
}

File: log-server/src/tests/config.test.ts
import test from 'node:test';
import assert from 'node:assert';
import { loadConfig, maxUploadBytes, retentionDurationMs, logDir } from '../config';

test('loadConfig defaults', () => {
  // Clear env vars
  const origPort = process.env.WUWAID_PORT;
  const origDataDir = process.env.WUWAID_DATA_DIR;
  const origMax = process.env.WUWAID_MAX_UPLOAD_MB;
  const origDays = process.env.WUWAID_RETENTION_DAYS;
  const origToken = process.env.WUWAID_ADMIN_TOKEN;

  delete process.env.WUWAID_PORT;
  delete process.env.WUWAID_DATA_DIR;
  delete process.env.WUWAID_MAX_UPLOAD_MB;
  delete process.env.WUWAID_RETENTION_DAYS;
  delete process.env.WUWAID_ADMIN_TOKEN;

  try {
    const cfg = loadConfig();
    assert.strictEqual(cfg.port, 8080);
    assert.ok(cfg.dataDir.length > 0);
    assert.strictEqual(cfg.maxUploadMB, 10);
    assert.strictEqual(cfg.retentionDays, 30);
    assert.strictEqual(cfg.adminToken, '');
  } finally {
    if (origPort) process.env.WUWAID_PORT = origPort;
    if (origDataDir) process.env.WUWAID_DATA_DIR = origDataDir;
    if (origMax) process.env.WUWAID_MAX_UPLOAD_MB = origMax;
    if (origDays) process.env.WUWAID_RETENTION_DAYS = origDays;
    if (origToken) process.env.WUWAID_ADMIN_TOKEN = origToken;
  }
});

test('loadConfig environment overrides', () => {
  const origPort = process.env.WUWAID_PORT;
  const origDataDir = process.env.WUWAID_DATA_DIR;
  const origMax = process.env.WUWAID_MAX_UPLOAD_MB;
  const origDays = process.env.WUWAID_RETENTION_DAYS;
  const origToken = process.env.WUWAID_ADMIN_TOKEN;

  process.env.WUWAID_PORT = '9090';
  process.env.WUWAID_DATA_DIR = '/tmp/wuwaid-test';
  process.env.WUWAID_MAX_UPLOAD_MB = '50';
  process.env.WUWAID_RETENTION_DAYS = '7';
  process.env.WUWAID_ADMIN_TOKEN = 'secret-token';

  try {
    const cfg = loadConfig();
    assert.strictEqual(cfg.port, 9090);
    assert.strictEqual(cfg.dataDir, '/tmp/wuwaid-test');
    assert.strictEqual(cfg.maxUploadMB, 50);
    assert.strictEqual(cfg.retentionDays, 7);
    assert.strictEqual(cfg.adminToken, 'secret-token');

    assert.strictEqual(maxUploadBytes(cfg), 50 * 1024 * 1024);
    assert.strictEqual(retentionDurationMs(cfg), 7 * 24 * 60 * 60 * 1000);
    assert.strictEqual(logDir(cfg), '/tmp/wuwaid-test/logs');
  } finally {
    if (origPort) process.env.WUWAID_PORT = origPort; else delete process.env.WUWAID_PORT;
    if (origDataDir) process.env.WUWAID_DATA_DIR = origDataDir; else delete process.env.WUWAID_DATA_DIR;
    if (origMax) process.env.WUWAID_MAX_UPLOAD_MB = origMax; else delete process.env.WUWAID_MAX_UPLOAD_MB;
    if (origDays) process.env.WUWAID_RETENTION_DAYS = origDays; else delete process.env.WUWAID_RETENTION_DAYS;
    if (origToken) process.env.WUWAID_ADMIN_TOKEN = origToken; else delete process.env.WUWAID_ADMIN_TOKEN;
  }
});
